src/ParseCallMap.hs

```diff
@@ -0,0 +1,313 @@
+  aForm (seqSubForm <|> choiceSubForm <|> callSubForm) <?> "call, choice or seq subform"
+  where
+    seqSubForm :: Parser (CallTree Ident)
+    seqSubForm = theIdentifier "seq" *> seqForest
+
+    callSubForm :: Parser (CallTree Ident)
+    callSubForm = Call <$ theIdentifier "call" <*> identForm
+
+    choiceSubForm :: Parser (CallTree Ident)
+    choiceSubForm = Choice <$ theIdentifier "choice" <*> subcallForm <*> subcallForm
+
+identForm :: Parser Ident
+identForm =
+  theForm "ident" (mkCIdent <$> (extractSpan <$> stringLiteral) <*> nodeForm)
+  where
+    mkCIdent s ni = CIdent.internalIdentAt (CPos.posOf ni) s -- FIXME: internalIdentAt isn't the right constructor
+
+nodeForm :: Parser CNode.NodeInfo
+nodeForm =
+  (CNode.undefNode <$ theIdentifier "noNode")
+  <|> (theForm "node" (mkCNodeInfo <$> cspanForm))
+
+data CSpan = NoPosCSpan | BuiltinPosCSpan | InternalPosCSpan | CSpanPosLength CPos.Position CPos.PosLength
+
+mkCNodeInfo :: CSpan -> CNode.NodeInfo
+mkCNodeInfo NoPosCSpan = CNode.mkNodeInfoOnlyPos CPos.nopos
+mkCNodeInfo BuiltinPosCSpan = CNode.mkNodeInfoOnlyPos CPos.builtinPos
+mkCNodeInfo InternalPosCSpan = CNode.mkNodeInfoOnlyPos CPos.internalPos
+mkCNodeInfo (CSpanPosLength pos poslen) = CNode.mkNodeInfoPosLen pos poslen
+
+cspanForm :: Parser CSpan
+cspanForm =
+  (NoPosCSpan <$ theIdentifier "noPos")
+  <|> (BuiltinPosCSpan <$ theIdentifier "builtin")
+  <|> (InternalPosCSpan <$ theIdentifier "internal")
+  <|> theForm "span" (mkSpanPos
+                       <$> sourceForm
+                       <*> (extractSpan <$> intLiteral)
+                       <*> sourceForm)
+  where
+    mkSpanPos pstart len pend = CSpanPosLength pstart (pend, len)
+
+sourceForm :: Parser CPos.Position
+sourceForm =
+  theForm "source" (mkPos
+                     <$> (extractSpan <$> stringLiteral)
+                     <*> (extractSpan <$> intLiteral)
+                     <*> (extractSpan <$> intLiteral)
+                     <*> (extractSpan <$> intLiteral))
+  where
+    mkPos path offset row column = CPos.position offset path row column
+
+------------------------------------------------------------
+-- Parsing S-expression files
+------------------------------------------------------------
+
+-- | A 'Span' is a contiguous region between two source positions.
+-- The first position in a span is the start the second is the end
+-- (we will maintain, but not check, the invariant that the start comes
+-- before the end in the file)
+type Span = (SourcePos, SourcePos)
+
+-- | A @Spanning a@ is a pair of a 'Span' and a value
+type Spanning a = (Span, a)
+
+-- | @Spanning a@ is a functor and this is its 'fmap'
+fmapSpan :: (a -> b) -> Spanning a -> Spanning b
+fmapSpan = fmap
+
+
+-- | @Spanning a@ is a comonad with 'extractSpan' the counit and 'extendSpan'
+-- its composition operations.
+extractSpan :: Spanning a -> a
+extractSpan = snd
+
+-- unused
+-- | @Spanning a@ is a comonad with 'extendSpan' its composition operation and
```

Comment:
The original history of this module was that it defined an intermediate representation of completely uninterpreted S-expressions and then traversed those trees to construct the `CallMap`.  So the S-expression nodes had `Spanning Sexp` children so I could give error messages with line numbers.

In the end I scrapped that whole approach because `Parsec`s 4-continuation error monad gives much better error messages than what I could cook up with `ExceptT` and because the code to just parse the forms I need directly is like half the size.

Which is all a roundabout way of saying that `Spanning` worked out great and I recommend something like it, but it's pretty extraneous here.